logs/ingress/src/audit_log.rs

```diff
@@ -0,0 +1,252 @@
+            user_id: None,
+        }
+    }
+
+    #[must_use]
+    fn max_duration(&self) -> Duration {
+        self.max_search_duration.unwrap_or(Duration::from_secs(15))
+    }
+
+    #[must_use]
+    fn make_backoff(&self) -> backoff::ExponentialBackoff {
+        let initial_interval = self
+            .initial_retry_interval
+            .unwrap_or(Duration::from_millis(400));
+        backoff::ExponentialBackoff {
+            max_elapsed_time: Some(self.max_duration()),
+            current_interval: initial_interval,
+            initial_interval: initial_interval,
+            max_interval: self.max_retry_interval.unwrap_or(Duration::from_secs(4)),
+            ..Default::default()
+        }
+    }
+}
+
+/// Possible errors while attempting to search for audit log entries
+pub enum Error {
+    SerenityError(serenity::Error),
+    Unauthorized(ErrorResponse),
+    SearchExhausted,
+    TimedOut,
+}
+
+type BackoffError<T> = backoff::Error<T>;
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct SearchTiming {
+    start: u64,
+    target: u64,
+}
+
+#[derive(Clone)]
+pub enum Strategy {
+    /// Uses the first audit log entry that is within the absolute bounds
+    /// and matches the supplied match function
+    First,
+    /// Creates a growing acceptance interval
+    GrowingInterval { max: Duration },
+}
+
+impl Strategy {
+    #[must_use]
+    fn matches<P>(
+        &self,
+        timing: SearchTiming,
+        entry: &AuditLogEntry,
+        search: &SearchQuery<P>,
+    ) -> bool
+    where
+        P: Fn(&AuditLogEntry) -> bool,
+    {
+        match self {
+            Self::First => true,
+            Self::GrowingInterval { max } => {
+                let timestamp = time::millisecond_ts();
+                // Construct the interval based on how much time has elapsed
+                // since the start
+                let ratio: f64 = max.as_secs_f64() / search.max_duration().as_secs_f64();
+                let ms_passed = timestamp - timing.start;
+                let interval_width = ((ms_passed as f64) * ratio) as u64;
+                let lower = timing.target - interval_width;
+                let upper = timing.target + interval_width;
+
+                // Only match if the entry's timestamp is inside the interval
+                let entry_ts = id::extract_timestamp(entry.id.0);
+                entry_ts > lower && entry_ts < upper
+            }
+        }
+    }
+}
+
+/// Determines if the given serenity error is an HTTP unauthorized
+fn unauthorized_response(err: &serenity::Error) -> Option<ErrorResponse> {
+    if let serenity::Error::Http(boxed_err) = err {
+        return match &**boxed_err {
+            serenity::http::HttpError::UnsuccessfulRequest(response) => {
+                if response.status_code == StatusCode::UNAUTHORIZED {
+                    return Some(response.clone());
+                } else {
+                    return None;
```

Comment:
you don't need these returns as you're returning the match block. just have the if and else evaluate to the Some(_) or None